### PR TITLE
Add functional regression metrics

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Add functional regression metrics. Discussed on slack.

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by rohitgr7** [comment]
Please let me know if something needs to be changed. Will write tests once current changes get approved.

**Comment by Borda** [comment]
so far it looks good to me :rabbit: 
cc: @justusschock @SkafteNicki

**Comment by SkafteNicki** [comment]
Changes until now also looks good to me, remember to update the functional docs with these new functions :)

**Comment by rohitgr7** [comment]
Yeah, will add tests and make changes to docs tonight✌

**Comment by rohitgr7** [comment]
Please let me know any other tests needed to be added or any doc updates.

**Comment by SkafteNicki** [comment]
@rohitgr7 PR looks great, if you could solve the merge conflicts (most related to the merge of #2483) I think this is ready to go :)

**Comment by rohitgr7** [comment]
@SkafteNicki Done :v:

File: pytorch_lightning/metrics/functional/regression.py [hunk 1]
```diff
@@ -4,6 +4,139 @@
...
+        target: ground truth labels
+        reduction: method for reducing mse (default: takes the mean)
+            Available reduction methods:
+
+            - elementwise_mean: takes the mean
+            - none: pass array
+            - sum: add elements
+
+    Return:
+        Tensor with mse
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/regression.py:26]:
```suggestion
        Tensor with MSE
```

File: pytorch_lightning/metrics/functional/regression.py [hunk 2]
```diff
@@ -4,6 +4,139 @@
...
+        target: ground truth labels
+        reduction: method for reducing rmse (default: takes the mean)
+            Available reduction methods:
+
+            - elementwise_mean: takes the mean
+            - none: pass array
+            - sum: add elements
+
+    Return:
+        Tensor with rmse
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/regression.py:60]:
```suggestion
        Tensor with RMSE
```

File: pytorch_lightning/metrics/functional/regression.py [hunk 3]
```diff
@@ -4,6 +4,139 @@
...
+        target: ground truth labels
+        reduction: method for reducing mae (default: takes the mean)
+            Available reduction methods:
+
+            - elementwise_mean: takes the mean
+            - none: pass array
+            - sum: add elements
+
+    Return:
+        Tensor with mae
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/regression.py:92]:
```suggestion
        Tensor with MAE
```

File: pytorch_lightning/metrics/functional/regression.py [hunk 4]
```diff
@@ -4,6 +4,139 @@
...
+        target: ground truth labels
+        reduction: method for reducing rmsle (default: takes the mean)
+            Available reduction methods:
+
+            - elementwise_mean: takes the mean
+            - none: pass array
+            - sum: add elements
+
+    Return:
+        Tensor with rmsle
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/regression.py:126]:
```suggestion
        Tensor with RMSLE
```

File: pytorch_lightning/metrics/functional/regression.py [hunk 5]
```diff
@@ -12,14 +145,22 @@ def psnr(
...
         base: a base of a logarithm to use (default: 10)
         reduction: method for reducing psnr (default: takes the mean)
+            Available reduction methods:
+
+            - elementwise_mean: takes the mean
+            - none: pass array
+            - sum add elements
+
+    Return:
+        Tensor with psnr score
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/regression.py:163]:
```suggestion
        Tensor with PSNR score
```

File: tests/metrics/functional/test_regression.py [hunk 6]
```diff
@@ -1,8 +1,89 @@
...
+    pytest.param(
+        [4., 3., 2., 1.],
+        [1., 4., 3., 2.],
+        ski_psnr(np.array([4., 3., 2., 1.]), np.array([1., 4., 3., 2.]), data_range=3)
+    )
+])
+def test_psnr(pred, target, expected):
+    score = psnr(pred=torch.tensor(pred),
+                 target=torch.tensor(target))
+    assert torch.allclose(score, torch.tensor(expected, dtype=torch.float), atol=1e-3)
```
Comment (Borda) [inline tests/metrics/functional/test_regression.py:66]:
```suggestion
@pytest.mark.parametrize(['pred', 'target'], [
    pytest.param([0., 1., 2., 3.], [0., 1., 2., 2.]),
    pytest.param([4., 3., 2., 1.], [1., 4., 3., 2.])
])
def test_psnr_with_skimage(pred, target, expected):
    score = psnr(pred=torch.tensor(pred),
                 target=torch.tensor(target))
    sk_score = ski_psnr(np.array(pred), np.array(target), data_range=3)
    assert torch.allclose(score, torch.tensor(sk_score, dtype=torch.float), atol=1e-3)
```

File: tests/metrics/functional/test_regression.py [hunk 7]
```diff
@@ -1,8 +1,89 @@
...
+        [1., 4., 3., 2.],
+        ski_psnr(np.array([4., 3., 2., 1.]), np.array([1., 4., 3., 2.]), data_range=4) * np.log(10)
+    )
+])
+def test_psnr_base_e_wider_range(pred, target, expected):
+    score = psnr(pred=torch.tensor(pred),
+                 target=torch.tensor(target),
+                 data_range=4,
+                 base=2.718281828459045)
+    assert torch.allclose(score, torch.tensor(expected, dtype=torch.float32), atol=1e-3)
```
Comment (Borda) [inline tests/metrics/functional/test_regression.py:86]:
```suggestion
@pytest.mark.parametrize(['pred', 'target'], [
    pytest.param([0., 1., 2., 3.], [0., 1., 2., 2.]),
    pytest.param([4., 3., 2., 1.], [1., 4., 3., 2.])
])
def test_psnr_base_e_wider_range(pred, target, expected):
    score = psnr(pred=torch.tensor(pred),
                 target=torch.tensor(target),
                 data_range=4,
                 base=2.718281828459045)
    sk_score = ski_psnr(np.array(pred), np.array(target), data_range=4) * np.log(10)
    assert torch.allclose(score, torch.tensor(sk_score, dtype=torch.float32), atol=1e-3)
```

File: tests/metrics/functional/test_regression.py [hunk 8]
```diff
@@ -12,26 +79,31 @@ def test_psnr_against_sklearn(sklearn_metric, torch_metric):
...
+
+    pred = torch.randint(10, (500,), device=device, dtype=torch.float)
+    target = torch.randint(5, (500,), device=device, dtype=torch.float)
+
+    sk_score = sklearn_metric(target.cpu().detach().numpy(),
+                              pred.cpu().detach().numpy(),
+                              data_range=5)
+    sk_score = torch.tensor(sk_score, dtype=torch.float, device=device)
+    th_score = torch_metric(pred, target, data_range=5)
+    assert torch.allclose(sk_score, th_score)
```
Comment (Borda) [inline tests/metrics/functional/test_regression.py:109]:
```suggestion
    for n_cls_pred, n_cls_target in [(10, 10), (5, 10), (10, 5)]:
        pred = torch.randint(n_cls_pred, (500,), device=device, dtype=torch.float)
        target = torch.randint(n_cls_target, (500,), device=device, dtype=torch.float)
    
        sk_score = sklearn_metric(target.cpu().detach().numpy(),
                                  pred.cpu().detach().numpy(),
                                  data_range=n_cls_target)
        sk_score = torch.tensor(sk_score, dtype=torch.float, device=device)
        pl_score = torch_metric(pred, target, data_range=n_cls_target)
        assert torch.allclose(sk_score, pl_score)
```

**Pull request merged**